### PR TITLE
Core: fix expression projection

### PR DESCRIPTION
**Pull request by chenjunjiedada**
This fixes partition filters on timestamp/date transform.

If we project the row filter on partition column, for example, `equal(day("timestamp1"), 100`, the projected predicate for date and timestamp transform should use the filed names in partition spec schema. Otherwise the evaluator will throw exception like below:
```
Cannot find field 'day' in struct: struct<1000: timestamp1_day: optional date, 1001: timestamp2_month: optional int, 1002: timestamp3_year: optional int, 1003: date1_day: optional date, 1004: date2_month: optional int, 1005: date3_year: optional int>
org.apache.iceberg.exceptions.ValidationException: Cannot find field 'day' in struct: struct<1000: timestamp1_day: optional date, 1001: timestamp2_month: optional int, 1002: timestamp3_year: optional int, 1003: date1_day: optional date, 1004: date2_month: optional int, 1005: date3_year: optional int>
	at org.apache.iceberg.exceptions.ValidationException.check(ValidationException.java:46)
	at org.apache.iceberg.expressions.NamedReference.bind(NamedReference.java:46)
	at org.apache.iceberg.expressions.NamedReference.bind(NamedReference.java:27)
	at org.apache.iceberg.expressions.UnboundPredicate.bind(UnboundPredicate.java:106)
	at org.apache.iceberg.expressions.Binder$BindVisitor.predicate(Binder.java:145)
	at org.apache.iceberg.expressions.Binder$BindVisitor.predicate(Binder.java:104)
	at org.apache.iceberg.expressions.ExpressionVisitors.visit(ExpressionVisitors.java:308)
	at org.apache.iceberg.expressions.Binder.bind(Binder.java:62)
...
```

### CODE REVIEW

**Review by Fokko — APPROVED** [review]

**Comment by rdblue** [comment]
@chenjunjiedada, can you add some context about what was broken? It looks like you're trying to fix the expression reference after projection. Was that causing an error while binding?

File: api/src/test/java/org/apache/iceberg/expressions/TestExpressionProjection.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+import static org.apache.iceberg.expressions.Expressions.day;
+import static org.apache.iceberg.expressions.Expressions.equal;
+import static org.apache.iceberg.expressions.Expressions.month;
+import static org.apache.iceberg.expressions.Expressions.year;
+import static org.apache.iceberg.types.Conversions.toByteBuffer;
+import static org.apache.iceberg.types.Types.NestedField.optional;
+import static org.apache.iceberg.types.Types.NestedField.required;
+
+public class TestExpressionProjection {
```
Comment (rdblue) [inline api/src/test/java/org/apache/iceberg/expressions/TestExpressionProjection.java:40]: Aren't there already tests for transform predicates? Can we update those to validate the field names instead?

Comment (chenjunjiedada) [reply api/src/test/java/org/apache/iceberg/expressions/TestExpressionProjection.java]: Make sense to me, I updated to test the predicate reference name instead.

**Comment by chenjunjiedada** [comment]
@rdblue, I just added some context in the issue description, also updated to test the filed name.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks for fixing this @chenjunjiedada! I merged it. The test failure was a flaky test in Java 8.

**Pull request merged**